In [1]:
import pandas as pd
import glob
import numpy as np
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
import matplotlib.pyplot as plt
%config InlineBackend.figure_format = 'retina'

In [2]:
!ls Transients_with_charging_plugin_info_05_02_to_23_09/

df_transients_charging_plugin_05 df_transients_charging_plugin_08
df_transients_charging_plugin_06 df_transients_charging_plugin_09
df_transients_charging_plugin_07


In [3]:
def load_pkl(path):
    filenames = glob.glob(path)

    dfs = []
    for filename in filenames:
        dfs.append(pd.read_pickle(filename))
    concat_df = pd.concat(dfs, ignore_index=True)
    return concat_df

In [4]:
def convert_to_datetime(timestamp_as_str):
    """Convert a timestring as str to pandas timestring dtype

    Args:
        timestamp_as_str: Timestamp in string format

    Returns:
        A readable timestamp for pandas
    """
    date, time, time_ns = timestamp_as_str.split('_')
    return pd.to_datetime(date + ' ' + time + '.' + time_ns, dayfirst=True)

In [5]:
data = load_pkl('Transients_with_charging_plugin_info_05_02_to_23_09/*')
data['begin_timestamp_string'] = data['begin_timestamp_string'].apply(convert_to_datetime)
data = data.sort_values('begin_timestamp_string')
data = data.drop(['filtered_signal', 'raw_signal_current', 'raw_signal_voltage', 'transient_rise_gradient'], axis=1)

In [6]:
data.head(10)

begin_timestamp_float_x  begin_index     begin_timestamp_string  phase_num  \
0             1.493604e+09         5604 2017-05-01 04:06:52.114772          2   
1             1.493640e+09          102 2017-05-01 13:55:40.753355          2   
2             1.493654e+09         5630 2017-05-01 17:52:10.525317          2   
3             1.493685e+09          647 2017-05-02 02:22:22.925197          2   
4             1.493698e+09         8750 2017-05-02 05:59:12.204228          2   
5             1.493698e+09         8754 2017-05-02 05:59:12.204237          3   
6             1.493700e+09         5403 2017-05-02 06:39:59.811533          2   
7             1.493700e+09         5404 2017-05-02 06:39:59.811535          3   
8             1.493700e+09         8750 2017-05-02 06:46:44.107575          1   
9             1.493700e+09         8750 2017-05-02 06:46:44.107575          2   

          three_first_peaks three_first_peaks_index  begin_timestamp_float_y  \
0    [25.251, 5.489, 2.660]      [5656, 5766, 5877]             1.493604e+09   
1    [27.796, 6.519, 0.948]        [134, 460, 8878]             1.493640e+09   
2    [25.127, 5.731, 2.847]      [5682, 5795, 5915]             1.493654e+09   
3    [25.011, 5.483, 2.721]         [699, 807, 915]             1.493685e+09   
4  [21.169, 12.606, 15.746]      [8757, 8771, 9443]             1.493698e+09   
5  [63.476, 16.495, 47.813]      [8767, 8781, 9437]             1.493698e+09   
6  [61.962, 20.032, 45.164]      [5416, 5431, 5956]             1.493700e+09   
7  [23.178, 14.446, 18.817]      [5413, 5428, 5953]             1.493700e+09   
8  [33.720, 13.077, 26.283]      [8756, 8773, 9272]             1.493700e+09   
9  [33.172, 11.947, 25.049]      [8757, 8772, 9273]             1.493700e+09   

   cars_total  one_phase  two_phases  three_phases  cars_phase_1  \
0         0.0        0.0         0.0           0.0           0.0   
1         0.0        0.0         0.0           0.0           0.0   
2         0.0        0.0         0.0           0.0           0.0   
3         0.0        0.0         0.0           0.0           0.0   
4         0.0        0.0         0.0           0.0           0.0   
5         0.0        0.0         0.0           0.0           0.0   
6         1.0        1.0         0.0           0.0           0.0   
7         1.0        1.0         0.0           0.0           0.0   
8         2.0        2.0         0.0           0.0           0.0   
9         2.0        2.0         0.0           0.0           0.0   

   cars_phase_2  cars_phase_3 plugin plugin_phase_1 plugin_phase_2  \
0           0.0           0.0  False          False          False   
1           0.0           0.0  False          False          False   
2           0.0           0.0  False          False          False   
3           0.0           0.0  False          False          False   
4           0.0           0.0   True          False          False   
5           0.0           0.0   True          False          False   
6           0.0           1.0  False          False           True   
7           0.0           1.0  False          False           True   
8           1.0           1.0  False          False          False   
9           1.0           1.0  False          False          False   

  plugin_phase_3  
0          False  
1          False  
2          False  
3          False  
4           True  
5           True  
6          False  
7          False  
8          False  
9          False

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4474 entries, 0 to 2644
Data columns (total 18 columns):
begin_timestamp_float_x    4474 non-null float64
begin_index                4474 non-null int64
begin_timestamp_string     4474 non-null datetime64[ns]
phase_num                  4474 non-null int64
three_first_peaks          4474 non-null object
three_first_peaks_index    4474 non-null object
begin_timestamp_float_y    4474 non-null float64
cars_total                 4474 non-null float64
one_phase                  4474 non-null float64
two_phases                 4474 non-null float64
three_phases               4474 non-null float64
cars_phase_1               4474 non-null float64
cars_phase_2               4474 non-null float64
cars_phase_3               4474 non-null float64
plugin                     4474 non-null object
plugin_phase_1             4474 non-null object
plugin_phase_2             4474 non-null object
plugin_phase_3             4474 non-null object
dtypes: dateti

# Clustering

Beschreibung der Cluster:

1: Transiente wenn kein Auto lädt oder angeschlossen wird

2: Transienten die während des Ladevorgangsauftreten, aber kein Plugin Event sind

3: Transienten die durch ein Plugin Event ausgelöst werden, dabei hat zuvor kein Auto geladen

4: Transienten die durch ein Plugin Event ausgelöst werden, allerdings mit zuvor geladenem Auto

In [12]:
def clustering_transients(row):
    if((row['cars_total'] == 0) & (not row['plugin'])):
        return 1
    else:
        if not (row['plugin']):
            return 2
        else:
            if(row['cars_total'] == 0):
                return 3
            else:
                return 4

In [15]:
%%timeit
data['cluster'] = data.apply(clustering_transients, axis=1)

394 ms ± 8.59 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [16]:
data.head(100)

begin_timestamp_float_x  begin_index     begin_timestamp_string  \
0              1.493604e+09         5604 2017-05-01 04:06:52.114772   
1              1.493640e+09          102 2017-05-01 13:55:40.753355   
2              1.493654e+09         5630 2017-05-01 17:52:10.525317   
3              1.493685e+09          647 2017-05-02 02:22:22.925197   
4              1.493698e+09         8750 2017-05-02 05:59:12.204228   
5              1.493698e+09         8754 2017-05-02 05:59:12.204237   
6              1.493700e+09         5403 2017-05-02 06:39:59.811533   
7              1.493700e+09         5404 2017-05-02 06:39:59.811535   
8              1.493700e+09         8750 2017-05-02 06:46:44.107575   
9              1.493700e+09         8750 2017-05-02 06:46:44.107575   
10             1.493700e+09         8750 2017-05-02 06:46:44.107575   
12             1.493701e+09         6106 2017-05-02 06:53:48.688569   
11             1.493701e+09         6107 2017-05-02 06:53:48.688571   
13             1.493701e+09         6107 2017-05-02 06:53:48.688571   
14             1.493704e+09         3762 2017-05-02 07:46:23.153438   
15             1.493706e+09          460 2017-05-02 08:14:54.429954   
16             1.493706e+09         5206 2017-05-02 08:15:34.456168   
17             1.493706e+09         5206 2017-05-02 08:15:34.456168   
18             1.493707e+09          702 2017-05-02 08:32:31.908947   
20             1.493707e+09         4756 2017-05-02 08:32:34.156612   
19             1.493707e+09         4757 2017-05-02 08:32:34.156614   
22             1.493708e+09          162 2017-05-02 08:57:23.262171   
21             1.493708e+09          163 2017-05-02 08:57:23.262172   
23             1.493708e+09          163 2017-05-02 08:57:23.262172   
24             1.493708e+09         3761 2017-05-02 08:59:28.950738   
25             1.493709e+09         8072 2017-05-02 09:16:00.232601   
26             1.493709e+09         8074 2017-05-02 09:16:00.232605   
27             1.493709e+09         8074 2017-05-02 09:16:00.232605   
29             1.493710e+09         6314 2017-05-02 09:30:42.589269   
30             1.493710e+09         6314 2017-05-02 09:30:42.589269   
..                      ...          ...                        ...   
70             1.493795e+09         7907 2017-05-03 09:05:38.897997   
71             1.493799e+09         3698 2017-05-03 10:02:45.035667   
72             1.493805e+09          604 2017-05-03 11:48:13.496216   
73             1.493805e+09         4381 2017-05-03 11:56:47.048507   
74             1.493808e+09         4568 2017-05-03 12:38:50.540181   
75             1.493808e+09         4568 2017-05-03 12:38:50.540181   
77             1.493811e+09         7422 2017-05-03 13:27:25.533053   
76             1.493811e+09         7423 2017-05-03 13:27:25.533055   
78             1.493815e+09          198 2017-05-03 14:29:58.421947   
79             1.493816e+09           55 2017-05-03 14:46:15.899351   
80             1.493816e+09         3544 2017-05-03 15:00:03.437346   
81             1.493816e+09         3544 2017-05-03 15:00:03.437346   
82             1.493816e+09         3544 2017-05-03 15:00:03.437346   
83             1.493816e+09         5971 2017-05-03 15:01:10.089678   
84             1.493816e+09         5972 2017-05-03 15:01:10.089680   
85             1.493817e+09         7137 2017-05-03 15:05:24.345771   
86             1.493818e+09         5152 2017-05-03 15:24:20.009839   
87             1.493830e+09          739 2017-05-03 18:41:48.837722   
88             1.493831e+09         9341 2017-05-03 18:55:25.029431   
89             1.493831e+09         9341 2017-05-03 18:55:25.029431   
90             1.493831e+09           93 2017-05-03 18:55:25.030945   
91             1.493860e+09          614 2017-05-04 03:00:52.024784   
92             1.493866e+09         1402 2017-05-04 04:41:30.215126   
93             1.493866e+09         1403 2017-05-04 04:41:30.215128   
94             1.493870e+09 

In [21]:
test.to_csv('test.csv')

# Einleitung
Für die Transientenanalyse werden Daten von Mai bis September untersucht. In diesem Transientendatensatz sind 4474 Events enthalten. Zunächst soll untersucht werden wie sich die Transienten verhalten, wenn ein Auto angeschlossen und wenn keins angeschlossen ist. 

## Hypothese 1
Wenn ein Auto zum Laden angeschlossen wird tretten stärkere und mehr Transienten auf.

In [ ]:
data['max_peak'] = data['three_first_peaks'].apply(lambda x: np.array(list(map(float, x))).max())

In [ ]:
data['car_charging'] = data['cars_total'].apply(lambda x: 1 if x > 0 else 0)

In [ ]:
charging_status = [data['car_charging'].value_counts()[0], data['car_charging'].value_counts()[1]]

# Set the bar labels
bar_labels = ['not charging', 'charging']

x_pos = list(range(len(bar_labels)))

plt.bar(x_pos, 
        charging_status, 
        align='center',
        color='#FFC222',
        alpha=0.5)

# set axes labels and title
plt.ylabel('Anzahl')
plt.xticks(x_pos, bar_labels)
plt.title('Transienten bei Auto nicht laden/laden')

plt.show()
print('Transient Events with no charging car: {}'.format(data['car_charging'].value_counts()[0]))
print('Transient Events with charging car: {}'.format(data['car_charging'].value_counts()[1]))


Hier ist zuerkennen, das sobald ein Auto lädt auch die Transienten zunehmen (um ca. 26%).

## Untersuchung des maximalen Spannungsspitzen von Transienten

In [ ]:
data.head()

## Transientenpeak für 1 phasen, 2 phasen und 3 phasen events unterscheiden

In [ ]:
data[data['three_phases'] > 0]

In [ ]:
data[data['car_charging'] == 0]['max_peak'].mean()

In [ ]:
np.std(data[data['car_charging'] == 0]['max_peak'])

In [ ]:
data[data['car_charging'] == 1]['max_peak'].mean()

In [ ]:
np.std(data[data['car_charging'] == 1]['max_peak'])

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
 
# Create data
N = 500
x = np.random.rand(N)
y = np.random.rand(N)
colors = (0,0,0)
area = np.pi*3
 
# Plot
plt.scatter(x, y, s=area, c=colors, alpha=0.5)
plt.title('Scatter plot pythonspot.com')
plt.xlabel('x')
plt.ylabel('y')
plt.show()

In [ ]:
# Create a list of the mean scores for each variable
mean_values = [df['pre_score'].mean(), df['mid_score'].mean(), df['post_score'].mean()]

# Create a list of variances, which are set at .25 above and below the score
variance = [df['pre_score'].mean() * 0.25, df['pre_score'].mean() * 0.25, df['pre_score'].mean() * 0.25]

# Set the bar labels
bar_labels = ['Pre Score', 'Mid Score', 'Post Score']

# Create the x position of the bars
x_pos = list(range(len(bar_labels)))

# Create the plot bars
# In x position
plt.bar(x_pos,
        # using the data from the mean_values
        mean_values, 
        # with a y-error lines set at variance
        yerr=variance, 
        # aligned in the center
        align='center',
        # with color
        color='#FFC222',
        # alpha 0.5
        alpha=0.5)

# add a grid
plt.grid()

# set height of the y-axis
max_y = max(zip(mean_values, variance)) # returns a tuple, here: (3, 5)
plt.ylim([0, (max_y[0] + max_y[1]) * 1.1])

# set axes labels and title
plt.ylabel('Score')
plt.xticks(x_pos, bar_labels)
plt.title('Mean Scores For Each Test')

plt.show()

In [ ]:
test = data['car_charging'].value_counts()

In [ ]:
mean_values = [df['pre_score'].mean(), df['mid_score'].mean(), df['post_score'].mean()]

In [ ]:
objects = ('Python', 'C++', 'Java', 'Perl', 'Scala', 'Lisp')
y_pos = np.arange(len(objects))
performance = [10,8,6,4,2,1]
 
plt.bar(y_pos, performance, align='center', alpha=0.5)
plt.xticks(y_pos, objects)
plt.ylabel('Usage')
plt.title('Programming language usage')
 
plt.show()

In [ ]:
data_car_not_charging = data[data['car_charging'] == 0]
data_car_charging = data[data['car_charging'] == 1]

In [ ]:
data_car_not_charging.info()

In [ ]:
mpg.head()

In [ ]:
from ggplot import *

ggplot(mpg, aes(x='cty')) + \
    geom_histogram() + \
    xlab("City MPG (Miles per Gallon)") + \
    ylab("# of Obs")

In [ ]:
%matplotlib inline

In [ ]:
iris = sns.load_dataset("iris")

In [ ]:
iris.head()

In [ ]:
import seaborn as sns
sns.set(style="darkgrid", color_codes=True)

#tips = sns.load_dataset("tips")
g = sns.jointplot("max_peak", "car_charging", data=data, kind="reg",
                  xlim=(0, 60), ylim=(0, 12), color="r", size=15)